packages/pulseaudio/module-sles-sink.c

```diff
@@ -125,23 +130,162 @@ static void sles_callback(SLBufferQueueItf bqPlayerBufferQueue, void *userdata)
+        [PA_CHANNEL_POSITION_FRONT_CENTER]          = SL_SPEAKER_FRONT_CENTER,
+        [PA_CHANNEL_POSITION_REAR_CENTER]           = SL_SPEAKER_BACK_CENTER,
+        [PA_CHANNEL_POSITION_REAR_LEFT]             = SL_SPEAKER_BACK_LEFT,
+        [PA_CHANNEL_POSITION_REAR_RIGHT]            = SL_SPEAKER_BACK_RIGHT,
+        [PA_CHANNEL_POSITION_LFE]                   = SL_SPEAKER_LOW_FREQUENCY,
+        [PA_CHANNEL_POSITION_FRONT_LEFT_OF_CENTER]  = SL_SPEAKER_FRONT_LEFT_OF_CENTER,
+        [PA_CHANNEL_POSITION_FRONT_RIGHT_OF_CENTER] = SL_SPEAKER_FRONT_RIGHT_OF_CENTER,
+        [PA_CHANNEL_POSITION_SIDE_LEFT]             = SL_SPEAKER_SIDE_LEFT,
+        [PA_CHANNEL_POSITION_SIDE_RIGHT]            = SL_SPEAKER_SIDE_RIGHT,
+        [PA_CHANNEL_POSITION_TOP_CENTER]            = SL_SPEAKER_TOP_CENTER,
+        [PA_CHANNEL_POSITION_TOP_FRONT_LEFT]        = SL_SPEAKER_TOP_FRONT_LEFT,
+        [PA_CHANNEL_POSITION_TOP_FRONT_RIGHT]       = SL_SPEAKER_TOP_FRONT_RIGHT,
+        [PA_CHANNEL_POSITION_TOP_FRONT_CENTER]      = SL_SPEAKER_TOP_FRONT_CENTER,
+        [PA_CHANNEL_POSITION_TOP_REAR_LEFT]         = SL_SPEAKER_TOP_BACK_LEFT,
+        [PA_CHANNEL_POSITION_TOP_REAR_RIGHT]        = SL_SPEAKER_TOP_BACK_RIGHT,
+        [PA_CHANNEL_POSITION_TOP_REAR_CENTER]       = SL_SPEAKER_TOP_BACK_CENTER,
+    };
+    pa_assert(x < PA_CHANNEL_POSITION_MAX);
+    return pa_channel_position_sl_speaker[x] ?: (SLuint32)(-1);
+}
+
+// sl_speaker_to_pa_channel_position converts an OpenSL ES speaker to the
+// equivalent PulseAudio channel position. If the channel does not have a PA
+// equivalent or is invalid, PA_CHANNEL_POSITION_INVALID is returned. Note that
+// this function does not handle the edge case where (SL_SPEAKER_FRONT_CENTER &&
+// channels == 1) is PA_CHANNEL_POSITION_MONO.
+//
+// https://freedesktop.org/software/pulseaudio/doxygen/channelmap_8h.html#af1cbe2738487c74f99e613779bd34bf2
+// https://www.khronos.org/files/opensl_es_1_0_provisional_specs.pdf#9.2.47
+static enum pa_channel_position sl_speaker_to_pa_channel_position(SLuint32 x) {
+    for (enum pa_channel_position pa = 0; pa < PA_CHANNEL_POSITION_MAX; pa++) {
+        if (pa != PA_CHANNEL_POSITION_MONO && pa_channel_position_to_sl_speaker(pa) == x) {
+            return pa;
+        }
+    }
+    return -1;
+}
+
+// pa_channel_map_to_sl_channel_mask converts a PulseAudio channel map to an
+// OpenSL ES channel mask. If an unknown or unsupported channel position is
+// found, -1 is returned. If map is NULL, 0 is returned.
+//
+// https://freedesktop.org/software/pulseaudio/doxygen/channelmap_8h.html#af1cbe2738487c74f99e613779bd34bf2
+// https://www.khronos.org/files/opensl_es_1_0_provisional_specs.pdf#9.1.8
+static SLuint32 pa_channel_map_to_sl_channel_mask(pa_channel_map *map) {
+    SLuint32 mask = 0, cur = 0, last = 0;
+    if (!map)
+        return 0;
+    for (int i = 0; i < map->channels; i++) {
+        if ((cur = pa_channel_position_to_sl_speaker(map->map[i])) == 0) {
```

Comment:
When/why would this return 0? Do you mean != -1?

Reply:
Oops, I need to fix that. I missed that during the final refactor.

Reply:
Fixed.